File: 백준/Silver/1743. 음식물 피하기/음식물 피하기.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

// 상하좌우 탐색해서 가장 큰 영역의 크기를 출력하면된다
// 첫번째줄에 rowN, colN, foodN
// 그다음줄부터 좌표가 주어지는데 -1 위치를 차감해서 넣어야한다
// grid에 좌표를 1로 마킹하고 시작하면된다
const solution = (input) => {
  const [rowN, colN, foodN] = input[0].split(" ").map((v) => Number(v));
  const grid = Array(rowN)
    .fill(0)
    .map((v) => Array(colN).fill(0));

  for (let i = 1; i <= foodN; i++) {
    const [x, y] = input[i].split(" ").map((v) => Number(v));
    grid[x - 1][y - 1] = 1;
  }

  let answer = 0;

  const dfs = (x, y) => {
    if (x < 0 || x >= rowN || y < 0 || y >= colN || grid[x][y] === 0) return 0;
    grid[x][y] = 0;

    const up = dfs(x - 1, y);
    const down = dfs(x + 1, y);
    const left = dfs(x, y - 1);
    const right = dfs(x, y + 1);

    return 1 + up + down + left + right;
  };

  for (let i = 0; i < rowN; i++) {
    for (let j = 0; j < colN; j++) {
      if (grid[i][j] === 1) {
        const area = dfs(i, j);
        answer = Math.max(answer, area);
      }
    }
  }

  console.log(answer);
};

solution(input);

File: 백준/Silver/1743. 음식물 피하기/음식물 피하기.py
import sys
sys.setrecursionlimit(10**6)

rowN, colN, n = list(map(int, input().split()))
grid = [[0]*colN for _ in range(rowN)]

for i in range(n):
  x, y = list(map(int, input().split()))
  grid[x-1][y-1] = 1

answer = 0 

def dfs(x, y):
  if x < 0 or x >= rowN or y < 0 or y >= colN or grid[x][y] == 0:
    return 0

  grid[x][y] = 0

  up = dfs(x-1, y)
  down = dfs(x+1, y)
  left = dfs(x, y-1)
  right = dfs(x, y+1)

  return 1 + up + down + left + right

for i in range(rowN):
  for j in  range(colN):
    if grid[i][j] == 1:
      answer = max(answer, dfs(i, j))

print(answer)